### PR TITLE
Add option to capture logback key value pairs

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/8059

### CODE REVIEW

File: instrumentation/logback/logback-appender-1.0/library/src/slf4j2ApiTest/resources/logback-test.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        %d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n
+      </pattern>
+    </encoder>
+  </appender>
+  <appender name="OpenTelemetry"
+    class="io.opentelemetry.instrumentation.logback.appender.v1_0.OpenTelemetryAppender">
+    <captureExperimentalAttributes>false</captureExperimentalAttributes>
+    <captureCodeAttributes>true</captureCodeAttributes>
+    <captureMarkerAttribute>true</captureMarkerAttribute>
+    <captureMdcAttributes>*</captureMdcAttributes>
```
Comment (jack-berg) [inline instrumentation/logback/logback-appender-1.0/library/src/slf4j2ApiTest/resources/logback-test.xml:16]: How come this doesn't need `<captureKeyValuePairAttributes>true</captureKeyValuePairAttributes>`?

Comment (laurit) [reply instrumentation/logback/logback-appender-1.0/library/src/slf4j2ApiTest/resources/logback-test.xml:17]: Thanks, fixed.

Comment (jack-berg) [reply instrumentation/logback/logback-appender-1.0/library/src/slf4j2ApiTest/resources/logback-test.xml:17]: Eventually, I hope we can make some / all of these enabled by default.

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/logback/logback-appender-1.0/library/src/slf4j2ApiTest/java/io/opentelemetry/instrumentation/logback/appender/v1_0/Slf4j2Test.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  @Test
+  void keyValue() {
+    logger.atInfo().setMessage("log message 1").addKeyValue("key", "value").log();
+
+    List<LogRecordData> logDataList = logRecordExporter.getFinishedLogItems();
+    assertThat(logDataList).hasSize(1);
+    LogRecordData logData = logDataList.get(0);
+    AssertionsForClassTypes.assertThat(logData.getResource()).isEqualTo(resource);
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/logback/logback-appender-1.0/library/src/slf4j2ApiTest/java/io/opentelemetry/instrumentation/logback/appender/v1_0/Slf4j2Test.java]: These were copied from `OpenTelemetryAppenderConfigTest`, fixed it here also.

File: instrumentation/logback/logback-appender-1.0/library/src/main/java/io/opentelemetry/instrumentation/logback/appender/v1_0/internal/LoggingEventMapper.java [hunk 3]
```diff
@@ -195,4 +202,77 @@ private static Severity levelToSeverity(Level level) {
...
+    }
+  }
+
+  @NoMuzzle
+  private static void captureMultipleMarkerAttributes(
+      AttributesBuilder attributes, ILoggingEvent loggingEvent) {
+    int i = 1;
+    for (Marker marker : loggingEvent.getMarkerList()) {
+      String markerName = marker.getName();
+      attributes.put(AttributeKey.stringKey("logback.marker." + i++), markerName);
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/logback/logback-appender-1.0/library/src/main/java/io/opentelemetry/instrumentation/logback/appender/v1_0/internal/LoggingEventMapper.java:259]: makes sense to me, may want to make `log4j.marker` attribute `string[]` also for consistency

**Review by trask — COMMENTED** [review]
> Not blocking: Would be interesting to see what happens with Slf4j using log4j as the backend. I think key value pairs show up in log4j context data. When adding log4j context data, we prefix all the keys with `log4j.context.`. So effectively, if you use SLF4j, the mapping experience differs based on whether you use log4j or logback as the backend.
>
> We should strive to make the two experiences consistent. 

it doesn't bother me too much at this point, since the instrumentation is at the logback/log4j layer (as opposed to at the slf4j layer), and we don't have any logging semantic conventions yet to unify the two experiences.

> Maybe we need a dedicated slf4j testing module where we apply the same abstract test to the logback and log4j implementations and verify the expected behavior?

👍 once we have logging semantic conventions to unify them

**Pull request merged**